nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/AccessResource.java

```diff
@@ -171,6 +204,516 @@ public Response getLoginConfig(@Context HttpServletRequest httpServletRequest) {
+        final String rawIdentity = samlCredential.getNameID().getValue();
+        final String mappedIdentity = IdentityMappingUtil.mapIdentity(rawIdentity, IdentityMappingUtil.getIdentityMappings(properties));
+        final long expiration = validateTokenExpiration(samlService.getAuthExpiration(), mappedIdentity);
+        final String issuer = samlCredential.getRemoteEntityID();
+
+        final LoginAuthenticationToken loginToken = new LoginAuthenticationToken(mappedIdentity, mappedIdentity, expiration, issuer);
+
+        // create and cache a NiFi JWT that can be retrieved later from the exchange end-point
+        samlStateManager.createJwt(samlRequestIdentifier, loginToken);
+
+        // store the SAMLCredential for retrieval during logout
+        // issue a delete first in case the user already had a stored credential that somehow wasn't properly cleaned up on logout
+        samlCredentialStore.delete(mappedIdentity);
+        samlCredentialStore.save(mappedIdentity, samlCredential);
+
+        // get the user's groups from the assertions if the exist and store them for later retrieval
+        final Set<String> userGroups = samlService.getUserGroups(samlCredential);
+        storeIdpGroups(mappedIdentity, IdpType.SAML, userGroups);
+
+        // redirect to the name page
+        httpServletResponse.sendRedirect(getNiFiUri());
+    }
+
+    private void storeIdpGroups(final String userIdentity, final IdpType idpType, final Set<String> groupNames) {
+        if (StringUtils.isBlank(userIdentity)) {
+            throw new IllegalArgumentException("User identity is required");
+        }
+
+        if (idpType == null) {
+            throw new IllegalArgumentException("IDP Type is required");
+        }
+
+        if (groupNames == null) {
+            throw new IllegalArgumentException("Group names are required");
+        }
+
+        // delete any previously saved groups
+        idpUserGroupService.deleteUserGroups(userIdentity);
+
+        if (groupNames.isEmpty()) {
+            logger.info("No groups returned from IDP for {}", userIdentity);
+            return;
+        }
+
+        final List<IdpUserGroup> idpUserGroups = new ArrayList<>();
+        final List<IdentityMapping> groupIdentityMappings = IdentityMappingUtil.getGroupMappings(properties);
+
+        for (final String groupName : groupNames) {
+            final String mappedGroupName = IdentityMappingUtil.mapIdentity(groupName, groupIdentityMappings);
+            logger.debug("Storing IDP groups, mapped {} to {}", groupName, mappedGroupName);
+
+            final IdpUserGroup idpUserGroup = new IdpUserGroup();
+            idpUserGroup.setIdentity(userIdentity);
+            idpUserGroup.setType(idpType);
+            idpUserGroup.setGroupName(mappedGroupName);
+            idpUserGroups.add(idpUserGroup);
+            logger.info("{} belongs to {}", userIdentity, mappedGroupName);
+        }
+
+        idpUserGroupService.createUserGroups(idpUserGroups);
```

Comment:
Does `deleteUserGroups` and `createUserGroups` need tobe atomic?

Reply:
Good question. I thought about this at one point and never came back to it. I suppose it might make make more sense to have a `replaceUserGroups` that could handle all of the logic within a single transaction.

In practice I'm not sure how much of a problem it would really be if the delete worked and then create failed, the failure on insert would fail the whole login process and they'd have to try again anyway.

I can work on this change though, I think its the more correct approach to be atomic.